dimod/constrained/constrained.py

```diff
@@ -1623,41 +1713,55 @@ def to_file(self, *,
-                     num_quadratic_variables_real=cqm.num_quadratic_variables('REAL', include_objective=True),
-                     num_linear_biases_real=cqm.num_biases('REAL', linear_only=True),
+                     num_quadratic_variables=cqm.num_quadratic_variables(include_objective=False),
+                     num_quadratic_variables_real=cqm.num_quadratic_variables(Vartype.REAL, include_objective=True),
+                     num_linear_biases_real=cqm.num_biases(Vartype.REAL, linear_only=True),
+                     num_weighted_constraints=sum(comp.lhs.is_soft() for comp in cqm.constraints.values()),
                      )
 
             it is terminated by a newline character and padded with spaces to
             make the entire length of the entire header divisible by 64.
 
             The constraint quadratic model data comes after the header. It is
-            encoded as a zip file. The zip file will contain one file
-            named `objective`, containing the objective as encoded as a file
-            view. It will also contain a directory called `constraints`. The
-            `constraints` directory will contain one subdirectory for each
-            constraint, each containing `lhs`, `rhs` and `sense` encoding
-            the `lhs` as a fileview, the `rhs` as a float and the sense
-            as a string. Each directory will also contain a `discrete` file,
-            encoding whether the constraint represents a discrete variable.
-            Weighted constraints also contain a `weight` and `penalty` file,
-            encoding the weight and the penalty type for the constraint.
-
-        Format Specification (Version 1.2):
-            This format is the same as Version 1.3, except that there are no
-            weighted constraints, and thus the `weight` and `penalty` files
-            are not present.
-
-        Format Specification (Version 1.1):
-
-            This format is the same as Version 1.2, except that the data dict
-            does not have ``num_quadratic_variables_real`` and
-            ``num_linear_biases_real``.
-
-        Format Specification (Version 1.0):
-
-            This format is the same as Version 1.1, except that the data dict
-            does not have `num_quadratic_variables`.
+            encoded as a zip file with the following structure
+
+            .. code-block:: bash
+
+                constraints/
+                    <label>/
+                        lhs
+                        rhs
+                        sense
+                        [discrete]
+                        [penalty]
+                        [weight]
+                    ...
+                objective
+                varinfo
+                [variable_labels.json]
+
+            The ``objective`` file encodes the objective.
+            See :meth:`~dimod.constrained.expression.ObjectiveView._to_file()`
+            for details about the file format.
```

Comment:
It would be nice to have the complete format spec decoupled from dimod internals.

Reply:
This is just deferring the documentation to another function so that it's easier to maintain. Though overall the documentation for this could/should be improved.